internal/querynode/impl.go

```diff
@@ -887,6 +889,40 @@ func (node *QueryNode) searchWithDmlChannel(ctx context.Context, req *querypb.Se
 	)
 	defer cancel()
 
+	// optimize search params
+	if req.Req.GetSerializedExprPlan() != nil && node.queryHook != nil {
+		segmentNum := len(cluster.segments) * channelNum
```

Comment:
`channelNum` is not accurate if it means the total channel number of one collection
also, this line accesses `cluster.segments` without mutex protection

Reply:
channelNum is got from len(req.GetDmlChannels()), how we get the  correct total channel number of one collection?

Reply:
`req.GetDmlChannels()` returns channel number in this request. If channels are loaded in different querynode, this length slice shall be 1. For now, shard leader does not contains information for shard number of one collection.

Reply:
add totalchannelnum to SearchRequest proto, help review again